sentry_flink/sentry_flink/flink/flink_adapter.py

```diff
@@ -113,15 +142,71 @@ def sink(self, step: Step, stream: Any) -> Any:
 
         return stream.sink_to(sink)
 
-    def map(self, step: Map, stream: Any) -> Any:
+    def filter(self, step: Filter, stream: DataStream) -> DataStream:
         imported_fn = self.load_function(step)
 
-        # TODO: Ensure output type is configurable like the schema above
-        return stream.map(func=lambda msg: imported_fn(msg), output_type=Types.STRING())
+        return stream.filter(func=lambda msg: imported_fn(msg))
 
-    def filter(self, step: Filter, stream: Any) -> Any:
+    def map(self, step: Map, stream: DataStream) -> DataStream:
         imported_fn = self.load_function(step)
-        return stream.filter(func=lambda msg: imported_fn(msg))
+
+        return_type = imported_fn.__annotations__["return"]
+
+        # TODO: Ensure output type is configurable like the schema above
+        return stream.map(
+            func=lambda msg: imported_fn(msg),
+            output_type=(
+                translate_to_flink_type(return_type)
+                if is_standard_type(return_type)
+                else translate_custom_type(return_type)
+            ),
+        )
+
+    def reduce(
+        self,
+        step: Reduce[MeasurementUnit, InputType, OutputType],
+        stream: DataStream,
+    ) -> DataStream:
+
+        agg = step.aggregate_fn
+        windowing = step.windowing
+
+        flink_window = build_flink_window(windowing)
+
+        # Optional parameters
+        group_by = step.group_by_key
+        agg_backend = step.aggregate_backend
+
+        # TODO: Configure WatermarkStrategy as part of KafkaSource
+        # Injecting strategy within a step like here produces
+        # a new, watermarked stream
+        watermark_strategy = WatermarkStrategy.for_monotonous_timestamps()
+        time_stream = stream.assign_timestamps_and_watermarks(watermark_strategy)
+
+        if group_by:
+            group_by_key = step.group_by_key
+            assert group_by_key is not None
+
+            keyed_stream = time_stream.key_by(FlinkGroupBy(group_by_key))
+
+            windowed_stream: Union[WindowedStream, AllWindowedStream] = keyed_stream.window(
+                flink_window
+            )
+
+        else:
+            windowed_stream = time_stream.window_all(flink_window)
+
+        return_type = agg().get_value.__annotations__["return"]
```

Comment:
Please consider adding a unit test for this method. 
You can mock the Flink environment and pass the mock to the initializer.

Reply:
I'm assuming you mean something different from sentry_flink/tests/test_pipeline.py?

Reply:
yes. 
test_pipeline tests very little about the result. Though I noticed it provides a lot more details regarding the reduce step. I think it may be ok. You can disregard this comment if you provide unit tests for the helper functions as per my previous comment